blueocean-events/src/main/java/io/jenkins/blueocean/events/BlueMessageEnricher.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package io.jenkins.blueocean.events;
+
+import hudson.Extension;
+import hudson.model.Item;
+import hudson.model.ItemGroup;
+import io.jenkins.blueocean.rest.hal.Link;
+import io.jenkins.blueocean.service.embedded.rest.OrganizationImpl;
+import io.jenkins.blueocean.service.embedded.rest.PipelineImpl;
+import jenkins.model.ParameterizedJobMixIn;
+import org.jenkins.pubsub.EventProps;
+import org.jenkins.pubsub.Events;
+import org.jenkins.pubsub.JobChannelMessage;
+import org.jenkins.pubsub.Message;
+import org.jenkins.pubsub.MessageEnricher;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject;
+
+import javax.annotation.Nonnull;
+
+/**
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
+@Extension
+public class BlueMessageEnricher extends MessageEnricher {
+    
+    enum BlueEventProps {
+        blueocean_job_rest_url,
+        blueocean_job_pipeline_name,
+        blueocean_job_branch_name,
+    }
+    
+    @Override
+    public void enrich(@Nonnull Message message) {
```

Comment:
@vivek Hey Vivek, I'm afraid that didn't work either. Same reason as before i.e. there's no active `StaplerRequest` instance so `ApiHead.INSTANCE()` throws an NPE.

```
org.jenkins.pubsub.MessageException: Event enrichment failure due to unexpected exception in io.jenkins.blueocean.events.BlueMessageEnricher.
	at org.jenkins.pubsub.PubsubBus.publish(PubsubBus.java:97)
	at org.jenkins.pubsub.listeners.SyncQueueListener.publish(SyncQueueListener.java:88)
	at org.jenkins.pubsub.listeners.SyncQueueListener.onLeft(SyncQueueListener.java:72)
	at hudson.model.Queue$LeftItem.enter(Queue.java:2609)
	at hudson.model.Queue.onStartExecuting(Queue.java:1126)
	at hudson.model.Executor$1.call(Executor.java:360)
	at hudson.model.Executor$1.call(Executor.java:346)
	at hudson.model.Queue._withLock(Queue.java:1371)
	at hudson.model.Queue.withLock(Queue.java:1232)
	at hudson.model.Executor.run(Executor.java:346)
Caused by: java.lang.NullPointerException
	at io.jenkins.blueocean.rest.ApiHead.INSTANCE(ApiHead.java:98)
	at io.jenkins.blueocean.service.embedded.rest.OrganizationImpl.getLink(OrganizationImpl.java:74)
	at io.jenkins.blueocean.service.embedded.rest.PipelineContainerImpl.getLink(PipelineContainerImpl.java:38)
	at io.jenkins.blueocean.service.embedded.rest.PipelineImpl$PipelineFactoryImpl.getPipeline(PipelineImpl.java:190)
	at io.jenkins.blueocean.service.embedded.rest.PipelineContainerImpl.get(PipelineContainerImpl.java:54)
	at io.jenkins.blueocean.service.embedded.rest.PipelineContainerImpl.get(PipelineContainerImpl.java:49)
	at io.jenkins.blueocean.events.BlueMessageEnricher.getLink(BlueMessageEnricher.java:85)
	at io.jenkins.blueocean.events.BlueMessageEnricher.enrich(BlueMessageEnricher.java:66)
	at org.jenkins.pubsub.PubsubBus.publish(PubsubBus.java:95)
	... 9 more
```